db/fetcher/fetcher.go

```diff
@@ -370,12 +516,59 @@ func (df *DocumentFetcher) FetchNext(ctx context.Context) (*encodedDocument, err
 			return nil, err
 		}
 
-		end, err := df.nextKey(ctx)
+		// check if we need to seek
+		seek := false
+		if df.filter != nil {
+			fmt.Println("checking for seek")
+			if df.filterSet.Equal(df.doc.filterSet) {
+				filterDoc, err := df.doc.decodeToDocForFilter()
+				if err != nil {
+					return nil, err
+				}
+
+				// run filter
+				passed, err := mapper.RunFilter(filterDoc, df.filter)
```

Comment:
suggestion: This at first glance looked like a horribly inefficient way of doing this, before I remembered that multiple fields might be involved. I think it is worth documenting this.

(also if for whatever reason you strongly disagree with that suggestion, please then remove the `// run filter` comment as it doesnt add any value - I assume you added it as you were planning out the implementation)

Reply:
there was a comment above the `if docDone` line that said `// if docDone` xD. Just my process, thanks for flagging the comment.

What seemed ineffecient at first glance? It gets run anyway, previously on scanner, now on fetcher.

Reply:
> What seemed ineffecient at first glance?

I honestly cant remember, will reply again if I think I remember and if this code is still relevant (just going through the old comments before reviewing the new code atm)